distributed/client.py

```diff
@@ -251,9 +251,10 @@ def type(self):
         return self._state.type
 
     def release(self):
-        if not self._cleared and self.client.generation == self._generation:
-            self._cleared = True
-            self.client._dec_ref(tokey(self.key))
+        with self.client._lock:
+            if not self._cleared and self.client.generation == self._generation:
+                self._cleared = True
+                self.client._dec_ref(tokey(self.key))
```

Comment:
After briefly looking at the code I think that we are safe here.  The client has much less state than the scheduler and is generally much easier to reason about.